src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -2905,6 +2908,187 @@ public void visitTypeCast(JCTypeCast tree) {
+                // Covers cases where the Type of the pattern is primitive e.g., v instanceof int
+                // - case type of v is ReferenceType, null check and unbox
+                // - case type of v is PrimitiveType
+
+                // rewrite instanceof if expr : wrapper reference type
+                //
+                // Integer v = ...
+                // if (v instanceof float)
+                // =>
+                // if (let tmp$123 = v; tmp$123 != null <&& if not unconditionally exact> ExactnessChecks.int_float(tmp$123.intValue()))
+                VarSymbol dollar_s = new VarSymbol(FINAL | SYNTHETIC,
+                        names.fromString("tmp" + tree.pos + this.target.syntheticNameChar()),
+                        tree.expr.type,
+                        currentMethodSym);
+
+                JCStatement var = make.at(tree.pos()).VarDef(dollar_s, instanceOfExpr).setType(dollar_s.type);
+
+                if (tree.expr.type.isReference()) {
+                    JCExpression nullCheck = makeBinary(NE,
+                            make.Ident(dollar_s),
+                            makeNull());
+
+                    if (types.checkUnconditionallyExact(types.unboxedType(tree.expr.type), tree.pattern.type)) {
+                        exactnessCheck = make.Literal(BOOLEAN, 1).setType(syms.booleanType);                                          // emit no exactness check
+                    } else {
+                        // if expression type is Byte, Short, Integer, ...
+                        // an unboxing conversion followed by a widening primitive conversion
+                        if (types.unboxedType(tree.expr.type).isPrimitive()) {
+                            exactnessCheck = getExactnessCheck(tree, boxIfNeeded(make.Ident(dollar_s), types.unboxedType(tree.expr.type))); // emit the exactness call
+                        } else {
+                            // if expression type is a supertype: Number, ..
+                            // a narrowing reference conversion followed by an unboxing conversion
+                            exactnessCheck = make.at(tree.pos()).TypeTest(tree.expr, make.Type(types.boxedClass(tree.pattern.type).type)).setType(syms.booleanType);;
+                        }
+                    }
+
+                    JCBinary nullCheckFollowedByExactnessCheckCall = makeBinary(AND,
+                            nullCheck,
+                            exactnessCheck);
+
+                    exactnessCheck = make.LetExpr(List.of(var), nullCheckFollowedByExactnessCheckCall)
+                            .setType(syms.booleanType);
+                } else {
+                    // rewrite instanceof if expr : primitive
+                    // int v = ...
+                    // if (v instanceof float)
+                    // =>
+                    // if (let tmp$123 = v; ExactnessChecks.int_float(tmp$123))
+                    JCIdent argument = make.Ident(dollar_s);
```

Comment:
In principle you don't need a synthetic variable here...

Reply:
Refactoring attempt No. 1 complete. I need to recheck the completeness of `checkUnconditionallyExact`. In a following commit.

Reply:
The new logic is much better and simpler to follow - thanks!